docs/set_env_vars.md
# Set environment variables


> Note: you don't need to set all of these variables.
> Docker compose files already have default values for most of them.
> But in production you have to set at least `DJANGO_SECRET_KEY`.
---
Also, when running / deploying microservices, there is some requirements:

> - use exact the same `DJANGO_SECRET_KEY` for all microservices;
> - all services except `core` needs to have `AUTH_DB_URL` set to `core` service's `DATABASE_URL`;
> - all services except `core` needs to have `BROKER_URL` set to `core` service's `BROKER_URL`.

Other requirements will be explained below.


---
## General

### Core service
To get started with core service on local machine, you don't need to set any environment variables.
To deploy it to the server in production mode, `DJANGO_SECRET_KEY`, 
`DJANGO_ALLOWED_HOSTS` and `DJANGO_CSRF_TRUSTED_ORIGINS` will be enough:

```dotenv
DJANGO_SECRET_KEY=<some-strict-secret-key>
```

### Fork app / microservice
To run second / third / etc. microservice somewhere, including server in production mode, 
you need to set at least the following variables:

```dotenv
# The same secret key as in core service
DJANGO_SECRET_KEY=<some-strict-secret-key>

# Any free port, for example 5433
POSTGRES_PORT=

# Any free port, for example 5051
PGADMIN_PORT=

# Some unique label, for example "products"
DJ_MS_APP_LABEL=

# Core service's database url
AUTH_DB_URL=postgres://postgres:postgres@host.docker.internal:5432/postgres

# Core service's broker url
BROKER_URL=amqp://rabbitmq:rabbitmq@host.docker.internal:5672

# Any free port, for example 8001
DJANGO_WEB_PORT=
```


---
## Django

```dotenv
DJANGO_DEBUG=
```

In production, `DJANGO_DEBUG` is set to `False` in the `docker-compose.prod.yml` file.
In other cases default value will be `True` if not set.
You don't need to set it here unless you want to change it to `False` in _development environment_.

<br>

```dotenv
DJANGO_SECRET_KEY=
```

Use 1 strict secret key for all microservices.
You can generate one using one of these commands:
- `openssl rand -base64 40`
- `openssl rand -hex 40`
- `python -c 'import uuid; print(uuid.uuid4().hex)'`

<br>


```dotenv
DJANGO_ALLOWED_HOSTS="<YOUR_HOSTNAME>,<ANOTHER_HOSTNAME>"
DJANGO_CSRF_TRUSTED_ORIGINS="<YOUR_HOSTNAME_WITH_HTTP_SCHEMA>,<ANOTHER_HOSTNAME>"
```

For example, you're deploying the project to the server with the hostname `example.com`.
Then you should set these variables to `example.com` and `https://example.com` respectively.


```dotenv
DJANGO_TIME_ZONE=
```

Default timezone is `UTC`. You can change it to your local timezone. For example: `Europe/Warsaw`

<br>

```dotenv
DJANGO_LANGUAGE_CODE=
```

Default language is English (`en-us`). You can change it to your local language. For example: `ru`

<br>

```dotenv
DJANGO_REST_AUTH_TOKEN_TTL=
```

This project has expiring token implementation. You can set the token expiration time in seconds.
Default value is `86400` seconds (24 hours).


---
## Database

> By default, the project uses postgres database, built in the `docker-compose.yml` file.
> So you don't need to set this setting unless you want to use another database.
> Default value is `postgres://postgres:postgres@postgres:5432/postgres`

```dotenv
DATABASE_URL=postgres://postgres:postgres@postgres:5432/postgres
```

### Postgres

```dotenv
POSTGRES_PORT=
```

Which port should be exposed for **postgres** that built in the `docker-compose.yml` file?
Default value is `5432`.

<br>

```dotenv
POSTGRES_USER=
```

Default **postgres** username is `postgres`.

<br>

```dotenv
POSTGRES_PASSWORD=
```

Default **postgres** password is `postgres`.

<br>

```dotenv
POSTGRES_DB=
```

Default **postgres** database is `postgres`.


---
## Microservices

```dotenv
DOCKER_BASE_IMAGE=
```

Which base image should be used for building the service? 
By default, `harleyking/dj-ms-core:latest` image is used.
This setting must be exactly the same in every microservice.

<br>

```dotenv
DJ_MS_APP_LABEL=
```

In main microservice this setting should be skipped.
In every other microservice you should set a unique label for the app.
It will be used for building appropriate docker image and as a url prefix.
For example: label `products` will result in the following docker image: `dj-ms-products:latest`.
And the url for the service will be: `http://localhost:8000/products/`

<br>

```dotenv
AUTH_DB_URL=
```

This is the most interesting part of the project! In core microservice this setting must be skipped.
But! In other microservices you should set this setting to the core microservice database url.
This allows you to use the same authentication database in all microservices.

<br>

```dotenv
BROKER_URL=
```

Same as `AUTH_DB_URL`, but for **RabbitMQ** message broker.
Skip this setting in the core microservice.
In other microservices set it to the core microservice broker url.

<br>

```dotenv
DJANGO_WEB_PORT=
```

Which http port should be exposed for entire service?
By default, `8000` port is used.
But if you're running multiple services on the same machine, you have to change it to another port.


---
## RabbitMQ

```dotenv
RABBITMQ_PORT=
```

Which port should be exposed for **RabbitMQ** that built in the `docker-compose.yml` file?
Default value is `5672`.

<br>

```dotenv
RABBITMQ_MANAGEMENT_PORT=
```

Which port should be exposed for **RabbitMQ** management that built in the `docker-compose.yml` file?
Default value is `15672`.

<br>

```dotenv
RABBITMQ_DEFAULT_USER=
```

Default username is `rabbitmq`.

<br>

```dotenv
RABBITMQ_DEFAULT_PASS=
```

Default password is `rabbitmq`.


---
### Sentry

```dotenv
SENTRY_DSN=
```

Your **Sentry** `DSN`. You can get it from your **Sentry** project settings.

<br>

```dotenv
SENTRY_ENVIRONMENT=
```

Default value is `development`.



core/settings.py
"""
Django settings for dj_ms_core project.

Generated by 'django-admin startproject' using Django 4.1.6.

For more information on this file, see
https://docs.djangoproject.com/en/4.1/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/4.1/ref/settings/
"""
import os
from pathlib import Path

import dj_database_url
import dotenv

# Build paths inside the project like this: BASE_DIR / 'subdir'.
BASE_DIR = Path(__file__).resolve().parent.parent


# Load env variables from *.env* file
dotenv_file = os.path.join(BASE_DIR, ".env")
if os.path.isfile(dotenv_file):
    dotenv.load_dotenv(dotenv_file)


# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/4.1/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = os.getenv('DJANGO_SECRET_KEY', 'django-insecure-9uq@=l5ykga+r$2=+nk+abwy)ej-y2n%!-u^fe^0=ybj)*+)d!')

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = False if os.getenv('DJANGO_DEBUG') == 'False' else True

# Use different app label for each microservice. For example, 'product' for product microservice
APP_LABEL = os.getenv('DJ_MS_APP_LABEL', '')

ALLOWED_HOSTS = [
    '127.0.0.1',
    'localhost'
]

extra_allowed_hosts = os.getenv('DJANGO_ALLOWED_HOSTS', None)

if extra_allowed_hosts:
    assert isinstance(extra_allowed_hosts.split(','), list)
    ALLOWED_HOSTS.extend(extra_allowed_hosts.split(','))

CORS_ORIGIN_ALLOW_ALL = False

CORS_ALLOWED_ORIGINS = (
    'http://127.0.0.1:8000',
    'http://localhost:8000',
)

CSRF_TRUSTED_ORIGINS = [
    'http://localhost:8000',
    'http://127.0.0.1:8000',
]

extra_csrf_trusted_origins = os.getenv('DJANGO_CSRF_TRUSTED_ORIGINS', None)

if extra_csrf_trusted_origins:
    assert isinstance(extra_csrf_trusted_origins.split(','), list)
    CSRF_TRUSTED_ORIGINS.extend(extra_csrf_trusted_origins.split(','))

# Application definition

INSTALLED_APPS = [
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',

    # third party apps
    'corsheaders',
    'rest_framework',
    'django_celery_beat',
    'django_celery_results',
    'auditlog',

    # microservice apps
    'authentication',
    'ms_auth_router',
    'app',

    # django_cleanup cleanup files after deleting model instance with FileField or ImageField fields
    'django_cleanup.apps.CleanupConfig'
]

MIDDLEWARE = [
    'core.middleware.HealthCheckMiddleware',
    'django.middleware.security.SecurityMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'corsheaders.middleware.CorsMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
    'auditlog.middleware.AuditlogMiddleware',
]

ROOT_URLCONF = 'core.urls'

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
        },
    },
]

WSGI_APPLICATION = 'core.wsgi.application'


# Database
# https://docs.djangoproject.com/en/4.1/ref/settings/#databases

DATABASE_ROUTERS = [
    'ms_auth_router.routers.DefaultRouter',
]

AUTH_DB = 'default'

DATABASES = {
    'default': dj_database_url.config(default='sqlite:///db.sqlite3', conn_max_age=600)
}

if auth_db := os.getenv('AUTH_DB_URL'):
    AUTH_DB = 'auth_db'
    DATABASES['auth_db'] = dj_database_url.parse(auth_db)

AUTH_USER_MODEL = 'authentication.User'

# -----> Rest Framework
REST_AUTH_TOKEN_MODEL = 'authentication.Token'

REST_AUTH_TOKEN_TTL = os.getenv('DJANGO_REST_AUTH_TOKEN_TTL', 60 * 60 * 24)

REST_AUTH_TOKEN_CREATOR = 'authentication.utils.create_token'

REST_FRAMEWORK = {
    'DEFAULT_AUTHENTICATION_CLASSES': [
        'rest_framework.authentication.SessionAuthentication',
        'authentication.utils.ExpiringTokenAuthentication'
    ],
    'DEFAULT_FILTER_BACKENDS': [
        'rest_framework.filters.SearchFilter',
        'rest_framework.filters.OrderingFilter',
        'django_filters.rest_framework.DjangoFilterBackend'
    ],
    'DATETIME_FORMAT': "%Y-%m-%d %H:%M:%S",
    'DEFAULT_PERMISSION_CLASSES': [
        'rest_framework.permissions.IsAuthenticated',
    ],
    'DEFAULT_PAGINATION_CLASS': 'core.api.pagination.PageNumberPagination',
    'PAGE_SIZE': 10
}


# Password validation
# https://docs.djangoproject.com/en/4.1/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
    },
]


# Internationalization
# https://docs.djangoproject.com/en/4.1/topics/i18n/

LANGUAGE_CODE = os.getenv('DJANGO_LANGUAGE_CODE', 'en-us')

TIME_ZONE = os.getenv('DJANGO_TIME_ZONE', 'UTC')

USE_I18N = True

USE_TZ = True


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/4.1/howto/static-files/

STATIC_URL = 'static/'

STATIC_ROOT = BASE_DIR / 'static'

STATICFILES_DIRS = [
    BASE_DIR / 'assets'
]

MEDIA_URL = 'media/'

MEDIA_ROOT = BASE_DIR / 'media'


# Default primary key field type
# https://docs.djangoproject.com/en/4.1/ref/settings/#default-auto-field

DEFAULT_AUTO_FIELD = 'django.db.models.BigAutoField'


# -----> RABBITMQ
BROKER_URL = os.getenv('BROKER_URL', 'amqp://guest:guest@localhost:5672')


# -----> CELERY
CELERY_BROKER_URL = BROKER_URL
CELERY_RESULT_BACKEND = 'django-db'
CELERY_CACHE_BACKEND = 'django-cache'
CELERY_ACCEPT_CONTENT = ['application/json']
CELERY_TASK_SERIALIZER = 'json'
CELERY_RESULT_SERIALIZER = 'json'
CELERY_TIMEZONE = TIME_ZONE
CELERY_TASK_DEFAULT_QUEUE = 'default'


# -----> SENTRY
if SENTRY_DSN := os.getenv('SENTRY_DSN', None):

    SENTRY_ENVIRONMENT = os.getenv('SENTRY_ENVIRONMENT', 'development')
    environment = f'{APP_LABEL}-{SENTRY_ENVIRONMENT}' if APP_LABEL else SENTRY_ENVIRONMENT

    import sentry_sdk
    from sentry_sdk.integrations.celery import CeleryIntegration
    from sentry_sdk.integrations.django import DjangoIntegration

    sentry_sdk.init(
        dsn=SENTRY_DSN,
        integrations=[
            DjangoIntegration(), CeleryIntegration()
        ],

        debug=DEBUG,

        # Set traces_sample_rate to 1.0 to capture 100%
        # of transactions for performance monitoring.
        # We recommend adjusting this value in production.
        traces_sample_rate=1.0,

        # If you wish to associate users to errors (assuming you are using
        # django.contrib.auth) you may enable sending PII data.
        send_default_pii=True,

        environment=environment
    )
